utils/open-isns/doc/isnsd.8
.TH ISNSD 8 "11 May 2007"
.SH NAME
isnsd \- iSNS server daemon
.SH SYNOPSIS
.B isnsd
.RB [ "\-f" ]
.RB [ "\-4" ]
.RB [ "\-6" ]
.RB [ "\-c \fIfilename" ]
.RB [ "\-d \fIdebug-facility" ]
.RB [ \--dump-db ]
.RB [ \--init ]

.SH DESCRIPTION
.B Isnsd
implements the iSNS protocol as defined in RFC 4171.
iSNS is a discovery protocol for iSCSI and iFCP.
.SH OPTIONS
By default,
.B isnsd
will take most of its settings from the configuration
file
.BR /etc/isns/isnsd.conf ,
with the exception of the following options:
.TP
.BI \--config " filename\fR, " \-c " filename
This option overrides the default configuration file.
.TP
.BR \--foreground , \-f
By default,
.B isnsd
will put itself into the background. By specifying this option, you can
tell it to run in the foreground. Any error messages or debug output
will be printed to the console rather than being sent to syslog.
.TP
.BI \-4
tells
.B isnsd
to create an IPv4 socket only. Normally, it defaults
to IPv6 (which will accept both IPv4 and IPv6 connections).
.TP
.BI \-6
tells
.B isnsd
explicitly
to create an IPv6 socket only. Since it defaults
to IPv6 anyway, this is really a no-op.
.TP
.BI \--debug " facility\fR, " \-d " facility
enables debugging. Valid facilities are
.PP
.TS
tab(,),box,center;
lb|lr.
socket,network send/receive
auth,authentication and security related information
message,iSNS protocol layer
state,database state
scn,SCN (state change notification) messages
esi,ESI (entity status inquiry) messages
all,all of the above
.TE
.PP
.TP
.B \--dump-db
This is a helper function that will read the database from the
file system, and display it in human readable form. When using
this option,
.B isnsd
will not open any sockets, and terminate immediately after display
the database.
.IP
This option is intended to be used by the administrator when suspecting
that the database contains bad/inconsistent information.
.TP
.B \--init
This option will create the server's authentication key, and
the required DSA parameters. The private key is stored in the
file specified by the
.B AuthKey
option (usually
.BR /etc/isns/auth_key ).
The public portion of the key is written to same directory,
with the suffix
.B .pub
appended to the key file name.
.SH SEE ALSO
RFC 4171,
.BR isnsadm (8),
.BR isnsdd (8),
.BR isns_config (5).
.SH AUTHORS
Olaf Kirch <olaf.kirch@oracle.com>


Changelog
open-iscsi-2.0-871 - open-iscsi-2.0.870

Boaz Harrosh (1):
      open-iscsi: Makefile: separate out user: and kernel: make targets

Erez Zilber (1):
      Add Module.markers to .gitignore

Karen Xie (2):
      userspace - add new transport cxgb3i
      userspace - setting interface ip address

Mike Christie (84):
      iscsid: fix relogin retry handling
      iscsid idbm: print out header and tail to seperate records
      ibft: add support to use iscsi_ibft module and log into all portals found in firmware
      fwparam: rm unused filepath argument
      iscsid idbm: move record strings to header
      iscsiadm: discovery mode supports the --show
      iscsiadm: have fw discovery handle --op arguments
      iscsiadm: fix output ordering
      iscsi modules: fix compilation
      iscsid/iscsiadm: fix strto* usage
      iscsid: fix logout response and time2wait handling
      iscsiadm: remove default bnx2i iface
      iscsid/iscsiadm: support multiple inititor names per host.
      iscsiadm: print session and session info running through ifaces in iface mode
      iscsid: update transport handle
      idbm: have idbm_rec_update_param update the value string of the matching rec_info
      iscsi mod: sync up kernel modules with 2.6.29
      iscsi mod: 2.6.27 compat
      iscsi mod: 2.6.26 compat
      iscsi mod: 2.6.24-25 compat
      iscsi mod: 2.6.20-23 compat
      iscsi mod: 2.6.14-19 compat
      iscsi mod: fix up 2.6.14-19 compat patch
      iscsi mod iscsi_tcp: compat fix for pI4/pI6
      build: add a make user
      iscsi tools: fix chap
      PATCH: iscsi tools: do not use exit()
      PATCH: fix iBFT firmware reading with newer kernels
      2.6.29-rc libiscsi: Fix scsi command timeout oops in iscsi_eh_timed_out
      iscsi mod: 2.6.14-19 compat
      iscsi tools: convert from strncat to strlcat
      iscsi tools: convert from strncpy to strlcpy
      iscsi tool make: add notification that iscsi start was made
      iscsid: Fix up connection failed messages
      docs: fix up iscsiadm man page iface info
      iscsi tools: update version in preparation for new release
      iscsid: flush sysfs cache.
      iscsid: mv event loop code to new file
      docs: update readme with libiscsi_tcp info




-------------------------------------------------------------------
open-iscsi-2.0-870 - open-iscsi-2.0.869

Doron Shoham (1):
      log.c: add error messages when allocation shared memory

Erez Zilber (2):
      Minor fixes in iscsi_discovery documentation
      rm unused variable in fw_entry.c

Hannes Reinecke (1):
      Add SLES10 SP2 compat

Hans de Goede (1):
      PATCH: add error checking to iscsi discovery db lock creation

Mike Christie (58):
      change mgmt_ipc to logout by sid
      iscsi class/if kernel: add ifacename attr
      add ifacename support tools
      rm db param
      rm num_transports from sysfs header.
      Add bind by initiatorname
      break up iface code and add default iser, tcp/default and bnx2i ifaces
      Use startup definition in util.c.
      set header digests to off by default
      fix compile warning for missing iface.h
      User 64 bit params mask.
      Do not allow iface setting to be changed in node mode.
      Fix bad merge. User iscsi param masks instead of params.
      Support mutlple ifaces with the same binding (lack of binding).
      Release transport entry when the transport is unloaded
      Fix ep_disconnect handling of invalid ep
      libiscsi: Fix nop timeout handling
      libiscsi: fix recv tmo
      Revert Release-transport-entry-when-the-transport-is-unload.patch
      pass ep to session creation
      Fix iser create bound session compat
      fix compilation on Fedora 9
      Fix sysfs handling of block:sdX and scsi bus changes
      Only autobind to ifaces with transport = tcp
      Increase login retry for iscsistart.
      Sync kernel modules to scsi-misc for 2.6.27.
      Update 2.6.14 - 2.6.19 compat patch
      Update 2.6.20 - 2.6.24 patch
      Update 2.6.24 - 2.6.25 patch
      iscsid: don't print enosys errors.
      libiscsi: support older tools that did not set can_queue/cmds_max
      Fix transport_name compat support.
      Fix idbm iscsid segfault when accsing ifaces
      Fix discovery and autobinding
      fix ipv6 login redirect support.
      Fix login redirect failure handling.
      remove sysfs_file
      fix dynamic tpgt support.
      Bump version for new release.
      Add compat patch for RHEL 5.2
      iscsid: adjust requested settings for user
      Sync kernel modules with 2.6.27
      Add 2.6.26 compat support
      Add Makefile support for 2.6.26 compat patch
      Add .gitignore files
      Revert broken SLES 10 compat patch.
      iscsi conf: increase default login max
      iscsi conf: partially revert increase default login max change
      modify initial login retry max
      libiscsi: fix data corruption when target has to resend data-in packets
      iscsi class: fix endpoint id handling
      handle ISCSI_ERR_INVALID_HOST
      iscsi_tcp: return a descriptive error value during connection errors
      libiscsi: fix locking in iscsi_eh_device_reset
      update 2.6.14-19_compat.patch
      update 2.6.20-21_compat.patch
      update 2.6.24_compat.patch
      Fix initiator.c compile warning



utils/open-isns/isnsd.c
/*
 * isnsd - the iSNS Daemon
 *
 * Copyright (C) 2007 Olaf Kirch <olaf.kirch@oracle.com>
 */

#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <unistd.h>

#ifdef MTRACE
# include <mcheck.h>
#endif

#include <isns.h>
#include "security.h"
#include "util.h"
#include "paths.h"
#include "internal.h"

enum {
	MODE_NORMAL,
	MODE_DUMP_DB,
	MODE_INIT,
};

static const char *	opt_configfile = ISNS_DEFAULT_ISNSD_CONFIG;
static int		opt_af = AF_UNSPEC;
static int		opt_mode = MODE_NORMAL;
static int		opt_foreground = 0;

static char *		slp_url;

static int		init_server(void);
static void		run_server(isns_server_t *, isns_db_t *);
static void		usage(int, const char *);
static void		cleanup(int);

static struct option	options[] = {
      {	"config",		required_argument,	NULL,	'c'		},
      {	"debug",		required_argument,	NULL,	'd'		},
      { "foreground",		no_argument,		NULL,	'f'		},
      { "init",			no_argument,		NULL,	MODE_INIT	},
      { "dump-db",		no_argument,		NULL,	MODE_DUMP_DB	},
      { "help",			no_argument,		NULL,	'h'		},
      { "version",		no_argument,		NULL,	'V'		},
      { NULL }
};

int
main(int argc, char **argv)
{
	isns_server_t	*server;
	isns_source_t	*source;
	isns_db_t	*db;
	int		c;

#ifdef MTRACE
	mtrace();
#endif

	while ((c = getopt_long(argc, argv, "46c:d:fh", options, NULL)) != -1) {
		switch (c) {
		case '4':
			opt_af = AF_INET;
			break;

		case '6':
			opt_af = AF_INET6;
			break;

		case 'c':
			opt_configfile = optarg;
			break;

		case 'd':
			isns_enable_debugging(optarg);
			break;

		case 'f':
			opt_foreground = 1;
			break;

		case MODE_DUMP_DB:
		case MODE_INIT:
			opt_mode = c;
			break;

		case 'h':
			usage(0, NULL);

		case 'V':
			printf("Open-iSNS version %s\n"
			       "Copyright (C) 2007, Olaf Kirch <olaf.kirch@oracle.com>\n",
			       OPENISNS_VERSION_STRING);
			return 0;

		default:
			usage(1, "Unknown option");
		}
	}

	if (optind != argc)
		usage(1, NULL);

	isns_read_config(opt_configfile);

	if (!isns_config.ic_source_name)
		usage(1, "Please specify an iSNS source name");
	source = isns_source_create_iscsi(isns_config.ic_source_name);

	if (opt_mode == MODE_INIT)
		return !init_server();

	if (opt_mode == MODE_NORMAL)
		isns_write_pidfile(isns_config.ic_pidfile);

	db = isns_db_open(isns_config.ic_database);
	if (db == NULL)
		isns_fatal("Unable to open database\n");

	if (opt_mode == MODE_DUMP_DB) {
		isns_db_print(db, isns_print_stdout);
		exit(0);
	}

	if (!opt_foreground) {
		if (daemon(0, 0) < 0)
			isns_fatal("Unable to background server process\n");
		isns_log_background();
		isns_update_pidfile(isns_config.ic_pidfile);
	}

	signal(SIGTERM, cleanup);
	signal(SIGINT, cleanup);

	server = isns_create_server(source, db, &isns_default_service_ops);

	run_server(server, db);
	return 0;
}

void
usage(int exval, const char *msg)
{
	if (msg)
		fprintf(stderr, "Error: %s\n", msg);
	fprintf(stderr,
	"Usage: isnsd [options]\n\n"
	"  --config        Specify alternative config fille\n"
	"  --foreground    Do not put daemon in the background\n"
	"  --debug         Enable debugging (list of debug flags)\n"
	"  --init          Initialize the server (key generation etc)\n"
	"  --dump-db       Display the database contents and exit\n"
	"  --help          Print this message\n"
	);
	exit(exval);
}

void
cleanup(int sig)
{
	isns_remove_pidfile(isns_config.ic_pidfile);
	exit(1);
}

static void
slp_cleanup(void)
{
	char	*url = slp_url;

	slp_url = NULL;
	if (url) {
		isns_slp_unregister(url);
		isns_free(url);
	}
}

/*
 * Initialize server
 */
int
init_server(void)
{
	if (!isns_security_init())
		return 0;

	/* Anything else? */

	return 1;
}

/*
 * Server main loop
 */
void
run_server(isns_server_t *server, isns_db_t *db)
{
	isns_socket_t	*sock;
	isns_security_t	*ctx = NULL;
	isns_message_t	*msg, *resp;
	int status;

	if (isns_config.ic_security) {
		const char	*ksname;
		isns_keystore_t	*ks;

		ctx = isns_default_security_context(1);
		if (!(ksname = isns_config.ic_client_keystore))
			isns_fatal("config problem: no key store specified\n");
		if (!strcasecmp(ksname, "db:"))
			ks = isns_create_db_keystore(db);
		else
			ks = isns_create_keystore(ksname);
		if (ks == NULL)
			isns_fatal("Unable to create keystore %s\n", ksname);
		isns_security_set_keystore(ctx, ks);
	}

	status = isns_dd_load_all(db);
	if (status != ISNS_SUCCESS)
		isns_fatal("Problem loading Discovery Domains from database\n");

	if (isns_config.ic_control_socket) {
		sock = isns_create_server_socket(isns_config.ic_control_socket,
				NULL, AF_UNSPEC, SOCK_STREAM);
		if (sock == NULL)
			isns_fatal("Unable to create control socket\n");
		/*
		isns_socket_set_security_ctx(sock, ctx);
		   */
	}

	sock = isns_create_server_socket(isns_config.ic_bind_address,
			"isns", opt_af, SOCK_STREAM);
	if (sock == NULL)
		isns_fatal("Unable to create server socket\n");
	isns_socket_set_security_ctx(sock, ctx);

	if (isns_config.ic_slp_register) {
		slp_url = isns_slp_build_url(0);
		isns_slp_register(slp_url);

		atexit(slp_cleanup);
	}

	isns_esi_init(server);
	isns_scn_init(server);

	while (1) {
		struct timeval timeout = { 0, 0 };
		time_t	now, then, next_timeout = time(NULL) + 3600;

		/* Expire entities that haven't seen any activity
		 * for a while. */
		if (isns_config.ic_registration_period) {
			then = isns_db_expire(db);
			if (then && then < next_timeout)
				next_timeout = then;
		}

		/* Run any timers (eg for ESI) */
		then = isns_run_timers();
		if (then && then < next_timeout)
			next_timeout = then;

		/* There may be pending SCNs, push them out now */
		then = isns_scn_transmit_all();
		if (then && then < next_timeout)
			next_timeout = then;

		/* Purge any objects that have been marked for removal
		 * from the DB (deleting them, or moving them to limbo
		 * state). */
		isns_db_purge(db);

		/* Determine how long we can sleep before working
		 * the ESI queues and DB expiry again. */
		now = time(NULL);
		if (next_timeout <= now)
			continue;
		timeout.tv_sec = next_timeout - now;

		if ((msg = isns_recv_message(&timeout)) == NULL)
			continue;

		if ((resp = isns_process_message(server, msg)) != NULL) {
			isns_socket_t *sock = isns_message_socket(msg);

			isns_socket_send(sock, resp);
			isns_message_release(resp);
		}

		isns_message_release(msg);
	}
}
